### PR TITLE
Add basic fault tolerance that requires replay from start.

### PR DESCRIPTION
**Pull request by blp**
This series adds durable input and output and uses them to build basic support for fault tolerance. This fault tolerance requires rerunning everything from the start of the input. Storage support, later, will allow restarting from a checkpoint.

I am really happy with how some commits in the middle of this series get rid of console spew running `cargo test -p dbsp_adapters`, properly capturing all the output.

This has inadequate testing, partly captured by issue #793.

Is this a user-visible change (yes/no): no

### CODE REVIEW

**Review by ryzhyk — APPROVED** [review]

File: crates/adapters/src/transport/durable_kafka/input.rs [hunk 1]
```diff
@@ -0,0 +1,1041 @@
...
+        write!(f, "Exit")
+    }
+}
+impl Error for Exit {}
+
+/// A reader for durable Kafka input.
+///
+/// Code outside this module accesses this as `dyn InputReader`.
+struct Reader {
+    action: Arc<Mutex<Action>>,
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:154]: I'd rather do this with an atomic! I didn't want to give up my `enum` value though, and Rust built-in atomics are just for basic types.

I see that there's a crate named `atomic` that might do what I want. I'll explore.

[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:154]: `atomic` won't work for this because my `enum` doesn't satisfy `NoUninit` from `bytemuck`.

Let's revisit this if it somehow shows up on a profile.

[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:154]: Oh, I see that crossbeam's `AtomicCell` is probably better. But my type is 16 bytes, so I think it would use a lock anyway.

File: crates/adapters/src/transport/durable_kafka/input.rs [hunk 2]
```diff
@@ -0,0 +1,1041 @@
...
+        let worker_thread = Self {
+            action: action.clone(),
+            commit_step: commit_step.clone(),
+            start_step,
+            parker,
+            config: config.clone(),
+            receiver: receiver.clone(),
+        };
+        Builder::new()
+            .name("worker(durable_Kafka)".into())
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:318]: Thread names are a problem because only the first 15 bytes show up on Linux. Everything after that gets hard-truncated.

I'll use the very short prefix `dkw-`.

File: crates/adapters/src/transport/durable_kafka/input.rs [hunk 3]
```diff
@@ -0,0 +1,1041 @@
...
+            .spawn(move || {
+                if let Err(error) = worker_thread.run() {
+                    if error.downcast_ref::<Exit>().is_some() {
+                        // Normal termination because of a requested exit.
+                    } else {
+                        info!("Durable Kafka input endpoint failed due to: {error:#}");
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:324]: I'm still traumatized from the Nicira/VMware QA policy that, if an ERROR shows up in the logs, a bug gets filed.

I'll change this to `error`.

File: crates/adapters/src/test/kafka.rs [hunk 4]
```diff
@@ -59,6 +60,14 @@ fn check_topics<C: ClientContext>(consumer: &Client<C>, topics: &[(&str, i32)])
 /// on drop.  Helps make sure that test runs don't leave garbage behind.
 impl KafkaResources {
     pub fn create_topics(topics: &[(&str, i32)]) -> Self {
+        // Kafka does not handle topic deletion and creation consistently when
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/test/kafka.rs:94]: I found it happening sometimes if there were multiple deletions and creations going on at the same time even if they were for different topics. I spent way too much time trying to track down the real issue, then I tried serializing and everything worked again.

File: crates/adapters/src/controller/mod.rs [hunk 5]
```diff
@@ -1146,6 +1173,28 @@ impl ControllerInner {
...
+    fn wait_to_settle(self: &Arc<Self>, step: Step) {
+        if !self.status.step_is_settled(step) {
+            let mut guard = self.unsettled_step.lock().unwrap();
+            loop {
+                let unsettled_step = self.status.unsettled_step().unwrap();
+                if unsettled_step > *guard {
+                    *guard = unsettled_step;
+                    self.step_settled.notify_all();
+                }
+                if unsettled_step < step {
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/controller/mod.rs:1190]: It's confusing because "unsettled_step" reverses the sense of what we care about. We want to wait for `step` to settle, which means that the first step that is *not* settled (unsettled) has to be beyond `step`, hence the condition.

The reason that we track the first unsettled step, rather than the last settled step, is that there might not be any settled steps at all, if we just initialized. We are left with at least the following choices, other than this one:
* Use `Option<Step>` to track the first settled step. It can be `None` before any steps settle. This is an annoying representation because of the special cases.
* Make the first step number 1 rather than number 0, and use 0 for the first settled step before any settle. This is viable and I'd be willing to do it this way.
* Use a signed type, e.g. `i64`, for `Step`, and use -1 for the first settled step before any settle. This is viable and I'd be willing to do it this way.

Basically, I'm not sure of the best representation for step numbers, and I'm happy to adopt a different representation at this point.

File: crates/adapters/src/transport/mod.rs [hunk 6]
```diff
@@ -157,21 +248,56 @@ pub trait InputEndpoint: Send {
 ///
 /// A transport endpoint pushes binary data downstream via an instance of this
 /// trait.
-// TODO: `input_owned`.
+///
+/// For a durable endpoint, where the data is divided into steps, there is some
+/// special terminology:
+///
+///   * "Committed" steps.  A step is "committed" when the endpoint has added
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/mod.rs:258]: That's better terminology. I'll do that.

File: crates/adapters/src/transport/mod.rs [hunk 7]
```diff
@@ -157,21 +248,56 @@ pub trait InputEndpoint: Send {
...
+    ///
+    /// A durable input transport keeps the order of fragments the same for a
+    /// given step from one read to the next.
     fn input_fragment(&mut self, data: &[u8]) -> Vec<ParseError>;
 
     /// Push a chunk of data to the consumer.
     ///
     /// The chunk is expected to contain complete records only.
+    ///
+    /// A durable input transport might shuffle the order of chunks for a given
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/mod.rs:293]: Hmm, that guarantee is already there, I just need to rewrite the comment, like this:

```rust
    /// Push a chunk of data to the consumer.
    ///
    /// The chunk is expected to contain complete records only.
    ///
    /// Some data in a durable input transport might not have an inherently
    /// defined order within a step.  The input endpoint may shuffle unordered
    /// chunks within a step from one read to the next.  For example, the
    /// durable Kafka reader will provide chunks from a given partition in the
    /// same order on each read, but it might interleave chunks from different
    /// partitions differently each time.
    fn input_chunk(&mut self, data: &[u8]) -> Vec<ParseError>;
```

**Review by mihaibudiu — APPROVED** [review]
Can't say I didn't get lost, but it's good to know what's going on

File: crates/adapters/src/controller/stats.rs [hunk 8]
```diff
@@ -565,6 +606,12 @@ pub struct InputEndpointMetrics {
     pub num_parse_errors: AtomicU64,
 
     pub end_of_input: AtomicBool,
+
+    /// The step to which arriving input belongs.
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/controller/stats.rs:610]: Done:

```patch
commit 738c4c033558bb07b47ea68e39e15b4570f8ebba
Author: Ben Pfaff <blp@feldera.com>
Date:   Tue Nov 7 14:40:46 2023 -0800

    dbsp_adapters: Better document step numbers.
    
    Signed-off-by: Ben Pfaff <blp@feldera.com>

diff --git a/crates/adapters/src/controller/stats.rs b/crates/adapters/src/controller/stats.rs
index aa2698f0..679eb601 100644
--- a/crates/adapters/src/controller/stats.rs
+++ b/crates/adapters/src/controller/stats.rs
@@ -31,7 +31,7 @@
 //! pending.
 
 use super::{EndpointId, InputEndpointConfig, OutputEndpointConfig, RuntimeConfig};
-use crate::{transport::Step, PipelineState};
+use crate::{transport::{Step, AtomicStep}, PipelineState};
 use anyhow::Error as AnyError;
 use crossbeam::sync::{ShardedLock, ShardedLockReadGuard, Unparker};
 use log::error;
@@ -608,10 +608,10 @@ pub struct InputEndpointMetrics {
     pub end_of_input: AtomicBool,
 
     /// The step to which arriving input belongs.
-    pub step: AtomicU64,
+    pub step: AtomicStep,
 
     /// The first step known not to have committed yet.
-    pub uncommitted: AtomicU64,
+    pub uncommitted: AtomicStep,
 }
 
 /// Input endpoint status information.
diff --git a/crates/adapters/src/transport/mod.rs b/crates/adapters/src/transport/mod.rs
index fe8a2128..a1a61716 100644
--- a/crates/adapters/src/transport/mod.rs
+++ b/crates/adapters/src/transport/mod.rs
@@ -68,6 +68,9 @@ pub mod durable_kafka;
 /// A "durable" data transport divides input into steps numbered sequentially.
 /// The first step is numbered zero.  Each step has durable content; that is, if
 /// it is read again, it will be the same as the first time.
+///
+/// The step number increases by 1 each time the circuit runs; that is, it
+/// tracks the global clock for the outermost circuit.
 pub type Step = u64;
 
 /// Atomic version of [`Step`].
```

> A question of terminology is if something belongs to a "step" or is labeled with a step.

I don't have a position on this grammatical issue.

File: crates/adapters/src/lib.rs [hunk 9]
```diff
@@ -111,6 +120,34 @@
...
+//! once, or to drop output or produce duplicate output, or to corrupt its
+//! internal state.
+//!
+//! The form of fault tolerance implemented in this crate can recover from
+//! crashes that kill processes or disrupt networking, but not crashes that lose
+//! storage or corrupt computations.
+//!
+//! Fault tolerance requires:
+//!
+//! * Input to be divided into numbered steps that can be retrieved repeatedly
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/lib.rs:138]: The circuit step() function gets called once per transport step, which is counted using `transport::Step`. I added some comments to clarify this.

What common abstraction do you see? I don't think that the circuit step() function counts global steps.

File: crates/adapters/src/transport/kafka/input.rs [hunk 10]
```diff
@@ -227,8 +189,106 @@ impl KafkaInputEndpointInner {
...
+                        // forward the error to upstream.
+                        let _ = consumer.input_chunk(payload);
+                    }
+                }
+                _ => (),
+            }
+
+            // Invalid broker address and other global errors are reported here.
+            if let Some((_error, reason)) = inner.pop_error() {
+                // let (_fatal, e) = endpoint.refine_error(error);
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/kafka/input.rs:263]: OK.

File: crates/adapters/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mihaibudiu) [reply crates/adapters/src/lib.rs:138]: it kindof does, for the outer nodes (not in loops).
I am expecting that the step count for input nodes in a DBPS graph is always the same as your step number.

File: crates/adapters/src/transport/kafka/mod.rs [hunk 11]
```diff
@@ -62,3 +80,103 @@ where
...
+/// `DeferredLogging` provides a way to capture log messages emitted by the unit
+/// tests.  A context can instantiate `DeferredLogging` and use
+/// `DeferredLogging::with_deferred_logging` to wrap calls to librdkafka
+/// functions that are likely to emit log messages.  This can be specific calls
+/// instead of every call, since we know what message the unit tests actually
+/// provoke.
+///
+/// When tests are disabled, this framework compiles to nothing.
+#[cfg(test)]
+pub(crate) struct DeferredLogging(Mutex<Option<Vec<(RDKafkaLogLevel, String, String)>>>);
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/kafka/mod.rs:104]: As written, it's Kafka-specific because it is only useful for implementing Kafka `ClientContext::log`.

File: crates/adapters/src/transport/durable_kafka/input.rs [hunk 12]
```diff
@@ -0,0 +1,1041 @@
...
+/// - `Ok(action)`: Run or pause.
+///
+/// - `Err(Exit)`: All done, please exit.
+///
+/// Representing an exit request as `Err` allows it to be implemented via `?`.
+type Action = Result<OkAction, Exit>;
+
+/// Error type to represent that the worker thread should exit.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+struct Exit;
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:142]: OK, I renamed it to ExitRequest.

File: crates/adapters/src/transport/durable_kafka/input.rs [hunk 13]
```diff
@@ -0,0 +1,1041 @@
...
+                    next_partition += 1;
+                    if next_partition >= n_partitions {
+                        next_partition = 0;
+                    }
+
+                    if lack_of_progress >= n_partitions {
+                        // Wait for at least one of these to happen:
+                        //
+                        //   - A message to arrive in one of the data partitions.
+                        //
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:591]: OK, I deleted them.

File: crates/adapters/src/transport/durable_kafka/input.rs [hunk 14]
```diff
@@ -0,0 +1,1041 @@
...
+/// Index reader for durable Kafka input and output.
+///
+/// The key to repeatability for Kafka input and output is to ensure that events
+/// in the data topic are always divided into steps in the same way.  We do this
+/// by tracking the division into steps in a separate "index" topic that
+/// consists of [`IndexEntry`] events, each of which maps a range of Kafka
+/// events in the data topic to a step number.
+///
+/// `Index` provides an interface for reading the index topic.
+struct Index<E>
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/transport/durable_kafka/input.rs:845]: OK, I renamed it to IndexReader.

File: crates/adapters/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/lib.rs:138]: Yes, I think that's true. If they count the steps, where do I get that count from?

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply crates/adapters/src/lib.rs:138]: I don't think that the circuit necessarily keeps track of this.
@ryzhyk?
But even if it doesn't, it would be useful to clarify that these 3 notions are the same one.

File: crates/adapters/src/controller/mod.rs [hunk 15]
```diff
@@ -276,6 +281,26 @@ impl Controller {
             .add_output_endpoint(endpoint_name, endpoint_config, endpoint)
     }
 
+    /// Reports whether the circuit is fault tolerant.  A circuit is fault
+    /// tolerant if it computes a deterministic function and all of its inputs
+    /// and outputs are durable.  This function assumes that the circuit is
+    /// deterministic.
+    pub fn is_fault_tolerant(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/controller/mod.rs:288]: That sounds like a good idea.

File: crates/adapters/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (blp) [reply crates/adapters/src/lib.rs:138]: They are the same. I added a clarifying comment.
